src/ls/ls.rs

```diff
@@ -0,0 +1,227 @@
+                Err(e)      => {
+                    show_error!("Cannot read directory '{}'. \n Reason: {}", loc, e);
+                    panic!();
+                },
+                Ok(entries) => enter_directory(entries, &options),
+            };
+        } 
+        
+        if p.is_file() {
+            display_item(Path::new(p), &options)
+        }
+    }
+}
+
+fn enter_directory(contents: ReadDir, options: &getopts::Matches) {
+    for entry in contents {
+        let entry = match entry {
+            Err(err) => {
+                show_error!("{}", err);
+                panic!();
+            },
+            Ok(en) => en
+        };
+        
+        // Currently have a DirEntry that we can believe in.
+        display_dir_entry(entry, options);
+    }    
+}
+
+fn display_dir_entry(entry: DirEntry, options: &getopts::Matches) {
+    let md = match entry.metadata() {
+        Err(e) => {
+            show_error!("Unable to retrieve metadata for {}. \n Error: {}", 
+                        display_file_name(entry.file_name()), e);
+            panic!();
+        },
+        Ok(md) => md
+    };
+    
+    println!(" {}{} {} somebody somegroup {: >9} {}",
+             display_file_type(entry.file_type()),
+             display_permissions(&md),
+             display_symlink_count(&md),
+             display_file_size(&md, options),
+             display_file_name(entry.file_name())
+            );
+}
+
+fn display_file_size(metadata: &Metadata, options: &getopts::Matches) -> String {
+    if options.opt_present("human-readable") {
+        convert(metadata.len() as f64)
+    } else {
+        metadata.len().to_string()
+    }
+}
+
+fn display_file_type(file_type: Result<FileType, std::io::Error>) -> String {
+    let file_type = match file_type {
+        Err(e) => {
+            show_error!("{}", e);
+            panic!()
+        },
+        Ok(ft) => ft 
+    };
+    
+    if file_type.is_dir() {
+        "d".to_string()
+    } else if file_type.is_symlink() {
+        "l".to_string()
+    } else {
+        "-".to_string()
+    }
+}
+
+fn display_file_name(name: OsString) -> String {
+    name.to_string_lossy().into_owned()
+}
+
+#[cfg(target_family = "windows")]
+#[allow(unused_variables)]
+fn display_symlink_count(metadata: &Metadata) -> String {
+    // Currently not sure of how to get this on Windows, so I'm punting.
+    // Git Bash looks like it may do the same thing.
+    String::from("1")
+}
+
+#[cfg(target_family = "unix")]
+fn display_symlink_count(metadata: &Metadata) -> String {
+    use std::os::unix::fs::MetadataExt;
+    
+    metadata.nlink().to_string()
+}
+
+#[cfg(target_family = "windows")]
+#[allow(unused_variables)]
+fn display_permissions(metadata: &Metadata) -> String {
+    String::from("---------")
+}
+
+#[cfg(target_family = "unix")]
+fn display_permissions(metadata: &Metadata) -> String {
+    use std::os::unix::fs::PermissionsExt;
+    
+    "xxxxxxxxx".to_string();
```

Comment:
Yeah, conditional compilation bit me on this. Latest commit removes this and a local VM will prevent it going forward. Apologies for the noise on the repo.